docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,221 @@
+The two reconstructed Jacobian matrices are then compared elementwise for equality.
+
+Slow real input numerical evaluation
+""""""""""""""""""""""""""""""""""""
+
+If we consider the elementary case of a one-dimensional function (:math:`N = M = 1`), then we can use the basic finite difference formula from `the wikipedia article <https://en.wikipedia.org/wiki/Finite_difference>`_. We use the "central difference" for better numerical properties:
+
+.. math::
+    \frac{\partial y}{\partial x} \approx \frac{f(x + eps) - f(x - eps)}{2 * eps}
+
+This formula easily generalizes for multiple outputs (:math:`M \gt 1`) by having :math:`\frac{\partial y}{\partial x}` be a column vector of size :math:`M \times 1` like :math:`f(x + eps)`.
+In that case, the above formula can be re-used as-is and approximates the full Jacobian matrix with only two evaluations of the user function (namely :math:`f(x + eps)` and :math:`f(x - eps)`).
+
+It is more computationally expensive to handle the case with multiple inputs (:math:`N \gt 1`). In this scenario, we loop over all the inputs one after the other and apply the :math:`eps` perturbation for each element of :math:`x` one after the other. This allows us to reconstruct the :math:`J_f` matrix column by column.
+
+Slow real input analytical evaluation
+"""""""""""""""""""""""""""""""""""""
+
+For the analytical evaluate, we use the fact, as described above, that backward mode AD computes :math:`v^T J_f`.
+For functions with a single output, we simply use :math:`v = 1` to recover the full Jacobian matrix with a single backward pass.
+
+For functions with more than one output, we resort to a for-loop which iterates over the outputs where each :math:`v` is a one-hot vector corresponding to each output one after the other. This allows to reconstruct the :math:`J_f` matrix row by row.
+
+Functions in :math:`\mathcal{C}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To test a function :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`, we reconstruct the (complex-valued) matrix that contains :math:`2 * CW`.
+
+Slow complex input numerical evaluation
+"""""""""""""""""""""""""""""""""""""""
+
+Consider the elementary case where :math:`N = M = 1` first. We know from (chapter 3 of) `this research paper <https://arxiv.org/pdf/1701.00392.pdf>`_) that:
+
+.. math::
+    CW := \frac{\partial y}{\partial z^*} = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})
+
+Note that :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}`, in the above equation, are :math:`\mathcal{R} \to \mathcal{R}` derivatives.
+To evaluate these numerically, we use the method described above for the real-to-real case.
+This allows us to compute the :math:`CW` matrix and then multiply it by :math:`2`.
+
+Note that the code, as of time of writing, computes this value in a slightly convoluted way:
```

Comment:
Out of curiosity, why do we compute it in a convoluted way?

Reply:
Mostly because the mathematical formulation that was derived a few months back when writing this was very different from the one here. And it lead to different formulas.

The two actually check the same thing in the end.